pkg/validation/compatibility.go

```diff
@@ -0,0 +1,153 @@
+package validation
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/blang/semver"
+	"io/ioutil"
+	"log"
+	"net/http"
+)
+
+type repository struct {
+	Id  string `json:"id"`
+	Url string `json:"url"`
+}
+
+type pluginMetadata struct {
+	PluginId    string    `json:"id"`
+	Description string    `json:"description"`
+	Provider    string    `json:"provider"`
+	Releases    []release `json:"releases"`
+}
+
+type release struct {
+	Version       string    `json:"version"`
+	Date          string    `json:"date"`
+	Requires      string    `json:"requires"`
+	Compatibility *platform `json:"compatibility"`
+	Hash          string    `json:"sha512sum"`
+	State         string    `json:"state"`
+	Url           string    `json:"url"`
+}
+type platform struct {
+	SpinnakerVersions []string `json:"spinnaker"`
+}
+
+type plugin struct {
+	Id      string
+	Version string
+}
+
+type CompatibilityResult struct {
+	PluginId      string
+	PluginVersion string
+	IsCompatible  bool
+	Reason        string
+}
+
+func ArePluginsCompatible(spinnakerVersion string, plugins []plugin, repo []string) []CompatibilityResult {
+	var result []CompatibilityResult
+	spinVersion, err := semver.Make(spinnakerVersion)
+	if err != nil {
+		log.Println(err)
+		return result
+	}
+	pluginsMetadata := getPluginMetadata(repo)
+	for _, v := range plugins {
+		comp, err := getCompatibilityConstraint(v.Id, v.Version, pluginsMetadata)
+		if err != nil {
+			log.Println(err)
+			result = append(result, CompatibilityResult{v.Id, v.Version, true, err.Error()})
+			continue
+		}
+		if comp == nil {
+			message := fmt.Sprintf("Plugin %s@%s does not contain compatibility constrain", v.Id, v.Version)
+			log.Printf(message)
+			result = append(result, CompatibilityResult{v.Id, v.Version, true, message})
+			continue
+		}
+		isCompatible := false
+		compatibleVersion := ""
+		for _, s := range comp.SpinnakerVersions {
+			compSpinVersion, err := semver.Make(s)
+			if err != nil {
+				log.Printf("%s in compatibility metadata is invalid", compSpinVersion)
+				continue
+			}
+			if spinVersion.Equals(compSpinVersion) || spinVersion.Minor == compSpinVersion.Minor {
+				isCompatible = true
+				compatibleVersion = s
+				break
+			}
+		}
+		if isCompatible {
+			result = append(result, CompatibilityResult{v.Id, v.Version, true, fmt.Sprintf("Plugin %s@%s compatible with version %s", v.Id, v.Version, compatibleVersion)})
+		} else {
+			result = append(result, CompatibilityResult{v.Id, v.Version, false, fmt.Sprintf("No compatible Spinnaker versions found for Plugin %s@%s", v.Id, v.Version)})
+		}
+	}
+	return result
+}
+
+func getPluginMetadata(repositories []string) []pluginMetadata {
```

Comment:
Same here: the return type should be `(error, []pluginMetadata)`. If the caller of this function (`ArePluginsCompatible`) sees an error, it should just immediately return the error.